### ISSUE TITLE
Language server exception in CodeAction handler

### ISSUE THREAD

**Issue by keyboardDrummer** [dafny-lang/dafny#2909]
When using the nightly Dafny LSP server, I get the following exception:

```
[Trace - 10:11:35 AM] Received response 'textDocument/codeAction - (38)' in 21ms. Request failed: Internal Error - System.Collections.Generic.KeyNotFoundException: The given key '74' was not present in the dictionary.
   at System.Collections.Generic.Dictionary`2.get_Item(TKey key)
   at Microsoft.Dafny.LanguageServer.Handlers.DafnyCodeActionInput.Extract(Range range)
   at Microsoft.Dafny.LanguageServer.Language.VerificationDafnyCodeActionProvider.GetDafnyCodeActions(IDafnyCodeActionInput input, Diagnostic diagnostic, Range selection)
   at Microsoft.Dafny.LanguageServer.Plugins.DiagnosticDafnyCodeActionProvider.GetDafnyCodeActions(IDafnyCodeActionInput input, Range selection)
   at Microsoft.Dafny.LanguageServer.Handlers.DafnyCodeActionHandler.<>c__DisplayClass5_0.<GetFixesWithIds>b__0(DafnyCodeActionProvider fixer)
   at System.Linq.Enumerable.SelectManySingleSelectorIterator`2.ToArray()
   at Microsoft.Dafny.LanguageServer.Handlers.DafnyCodeActionHandler.Handle(CodeActionParams request, CancellationToken cancellationToken)
   at OmniSharp.Extensions.LanguageServer.Server.Pipelines.SemanticTokensDeltaPipeline`2.Handle(TRequest request, CancellationToken cancellationToken, RequestHandlerDelegate`1 next)
   at OmniSharp.Extensions.LanguageServer.Server.Pipelines.ResolveCommandPipeline`2.Handle(TRequest request, CancellationToken cancellationToken, RequestHandlerDelegate`1 next)
   at MediatR.Pipeline.RequestPreProcessorBehavior`2.Handle(TRequest request, CancellationToken cancellationToken, RequestHandlerDelegate`1 next)
   at MediatR.Pipeline.RequestPostProcessorBehavior`2.Handle(TRequest request, CancellationToken cancellationToken, RequestHandlerDelegate`1 next)
   at MediatR.Pipeline.RequestExceptionProcessorBehavior`2.Handle(TRequest request, CancellationToken cancellationToken, RequestHandlerDelegate`1 next)
   at MediatR.Pipeline.RequestExceptionProcessorBehavior`2.Handle(TRequest request, CancellationToken cancellationToken, RequestHandlerDelegate`1 next)
   at MediatR.Pipeline.RequestExceptionActionProcessorBehavior`2.Handle(TRequest request, CancellationToken cancellationToken, RequestHandlerDelegate`1 next)
   at MediatR.Pipeline.RequestExceptionActionProcessorBehavior`2.Handle(TRequest request, CancellationToken cancellationToken, RequestHandlerDelegate`1 next)
   at OmniSharp.Extensions.JsonRpc.RequestRouterBase`1.<RouteRequest>g__InnerRoute|7_0(IServiceScopeFactory serviceScopeFactory, Request request, TDescriptor descriptor, Object params, CancellationToken token, ILogger logger)
   at OmniSharp.Extensions.JsonRpc.RequestRouterBase`1.RouteRequest(IRequestDescriptor`1 descriptors, Request request, CancellationToken token)
   at OmniSharp.Extensions.JsonRpc.DefaultRequestInvoker.<>c__DisplayClass10_0.<<RouteRequest>b__5>d.MoveNext() (-32603).
```

On the program:

```
class {:extern} NativeString {

  ghost var value: string

  static method {:extern} Create(dafnyString: string) returns (r: NativeString)
    ensures r.value == dafnyString

  function method {:extern} Concat(other: NativeString): NativeString
    ensures value + other.value == Concat(other).value

  // returns something that lazily decodes the underlying native string to Dafny chars, as the LinkedList is unfolded using Tail
  method {:extern} Iterate() returns (r: LinkedList<char>)
    ensures r.value == value
}

trait LinkedList<T> {
  ghost var value: seq<T>

  predicate method IsCons() 
    ensures IsCons() == (|value| > 0)

  function Head(): T 
    requires IsCons()
    ensures Head() == value[0]

  method Tail() returns (r: LinkedList<T>) 
    requires |value| > 0
    ensures r.value == value[1..]
}

class {:extern} NativeStringIterator extends LinkedList<char> {

  predicate method {:extern} IsCons()
    ensures IsCons() == (|value| > 0)

  function {:extern} Head(): char
    requires IsCons()
    ensures Head() == value[0]

  method {:extern} Tail() returns (r: LinkedList<char>)
    requires IsCons()
    ensures r.value == value[1..]
}

method HasPrefix(ns: NativeString, prefix: string) returns (r: bool) 
  ensures r == (|ns.value| >= |prefix| ==> (ns.value[..|prefix|] == prefix)) {
  var iter := ns.Iterate();
}
```

**Comment by MikaelMayer** [comment]
I can't reproduce it  It has something to do with the fact that the buffer is not in sync with the lines. Could you please tell me a precise repro?

**Comment by franck44** [comment]
I may experience a related issue with the language server (Dafny 3.9.0, MacOS) in VSCode.

Sequence of events is:

1. open a file
2. add some chars/lines
3.  "resolving" starts and never ends 

Every time I add some text to a file (without saving, and my settings in VScode are "verify on save"), the "resolution/parsing" seems to be triggered: "resolving" appears in the status bar and never terminates, and finally generates an error message:



```
[Error - 09:02:32] Request textDocument/documentSymbol failed.
  Message: Internal Error - System.NullReferenceException: Object reference not set to an instance of an object.
   at Microsoft.Dafny.Resolver.SeqTypeArgToBeResolved.ToString()
   at Microsoft.Dafny.TypeConstraint.ErrorMsg.<Reporting>g__RemoveAmbiguity|5_0(Object[] msgArgs)
   at Microsoft.Dafny.TypeConstraint.ErrorMsg.Reporting(ErrorReporter reporter, String suffix)
   at Microsoft.Dafny.TypeConstraint.ReportErrors(ErrorReporter reporter)
   at Microsoft.Dafny.Resolver.SolveAllTypeConstraints()
   at Microsoft.Dafny.Resolver.ResolveTopLevelDecls_Core(List`1 declarations, Graph`1 datatypeDependencies, Graph`1 codatatypeDependencies, Boolean isAnExport)
   at Microsoft.Dafny.Resolver.ResolveProgram(Program prog)
   at Microsoft.Dafny.LanguageServer.Language.Symbols.DafnyLangSymbolResolver.RunDafnyResolver(TextDocumentItem document, Program program)
   at Microsoft.Dafny.LanguageServer.Language.Symbols.DafnyLangSymbolResolver.ResolveSymbols(TextDocumentItem textDocument, Program program, Boolean& canDoVerification, CancellationToken cancellationToken)
   at Microsoft.Dafny.LanguageServer.Workspace.TextDocumentLoader.LoadInternal(DocumentTextBuffer textDocument, CancellationToken cancellationToken)
   at Microsoft.Dafny.LanguageServer.Workspace.TextDocumentLoader.<>c__DisplayClass12_0.<<LoadAsync>b__0>d.MoveNext()
--- End of stack trace from previous location ---
   at Microsoft.Dafny.LanguageServer.Workspace.TextDocumentLoader.LoadAsync(DocumentTextBuffer textDocument, CancellationToken cancellationToken)
   at Microsoft.Dafny.LanguageServer.Workspace.Compilation.ResolveAsync()
   at Microsoft.Dafny.LanguageServer.Workspace.DocumentManager.GetSnapshotAfterResolutionAsync()
   at Microsoft.Dafny.LanguageServer.Handlers.DafnyDocumentSymbolHandler.Handle(DocumentSymbolParams request, CancellationToken cancellationToken)
   at OmniSharp.Extensions.LanguageServer.Server.Pipelines.SemanticTokensDeltaPipeline`2.Handle(TRequest request, CancellationToken cancellationToken, RequestHandlerDelegate`1 next)
   at OmniSharp.Extensions.LanguageServer.Server.Pipelines.ResolveCommandPipeline`2.Handle(TRequest request, CancellationToken cancellationToken, RequestHandlerDelegate`1 next)
   at MediatR.Pipeline.RequestPreProcessorBehavior`2.Handle(TRequest request, CancellationToken cancellationToken, RequestHandlerDelegate`1 next)
   at MediatR.Pipeline.RequestPostProcessorBehavior`2.Handle(TRequest request, CancellationToken cancellationToken, RequestHandlerDelegate`1 next)
   at MediatR.Pipeline.RequestExceptionProcessorBehavior`2.Handle(TRequest request, CancellationToken cancellationToken, RequestHandlerDelegate`1 next)
   at MediatR.Pipeline.RequestExceptionProcessorBehavior`2.Handle(TRequest request, CancellationToken cancellationToken, RequestHandlerDelegate`1 next)
   at MediatR.Pipeline.RequestExceptionActionProcessorBehavior`2.Handle(TRequest request, CancellationToken cancellationToken, RequestHandlerDelegate`1 next)
   at MediatR.Pipeline.RequestExceptionActionProcessorBehavior`2.Handle(TRequest request, CancellationToken cancellationToken, RequestHandlerDelegate`1 next)
   at OmniSharp.Extensions.JsonRpc.RequestRouterBase`1.<RouteRequest>g__InnerRoute|7_0(IServiceScopeFactory serviceScopeFactory, Request request, TDescriptor descriptor, Object params, CancellationToken token, ILogger logger)
   at OmniSharp.Extensions.JsonRpc.RequestRouterBase`1.RouteRequest(IRequestDescriptor`1 descriptors, Request request, CancellationToken token)
   at OmniSharp.Extensions.JsonRpc.DefaultRequestInvoker.<>c__DisplayClass10_0.<<RouteRequest>b__5>d.MoveNext()
  Code: -32603 
```

If I "reload window" the file is correctly parsed and verified.

**Comment by atomb** [comment]
This is a duplicate of #2931 and #2828, and all three should be resolved with the fix of #2828.
